### PR TITLE
CI/CD + go v1.19

### PR DESCRIPTION
**Pull request by JoshVanL**
PR makes a number of changes:

- Update go to use `v1.19`
- Updates the go workflow to use Google identity federation
- Changes the e2e tests to use the credentials sourced from the GitHub workflow

### CODE REVIEW

File: .github/workflows/tests.yml [hunk 1]
```diff
@@ -12,20 +12,35 @@ on:
...
+    - name: Run e2e tests
+      run: make e2e
+      env:
+        TEST_GOOGLE_APPLICATION_CREDENTIALS: ${{ steps.auth.outputs.credentials_file_path }}
+
+    - uses: actions/upload-artifact@v3
+      if: always()
+      with:
+        name: e2e-logs
+        path: _artifacts/e2e/logs
```
Comment (JoshVanL) [inline .github/workflows/tests.yml:46]: Google identity federation. Pass the json blob to the test

Comment (JoshVanL) [reply .github/workflows/tests.yml:46]: Upload e2e test logs as a GA artifact

File: Makefile [hunk 2]
```diff
@@ -1,39 +1,25 @@
...
-	$(KUBECTL) --kubeconfig kubeconfig.yaml apply -f https://github.com/jetstack/cert-manager/releases/download/v1.5.1/cert-manager.yaml
-	$(KUSTOMIZE) build config/crd | $(KUBECTL) --kubeconfig kubeconfig.yaml apply -f -
-	cd config/manager; $(KUSTOMIZE) edit set image controller=${IMG}
-	$(KUSTOMIZE) build config/default | $(KUBECTL) --kubeconfig kubeconfig.yaml apply -f -
-	timeout 5m bash -c 'until $(KUBECTL) --kubeconfig kubeconfig.yaml --timeout=120s wait --for=condition=Ready pods --all --namespace kube-system; do sleep 1; done'
-	timeout 5m bash -c 'until $(KUBECTL) --kubeconfig kubeconfig.yaml --timeout=120s wait --for=condition=Ready pods --all --namespace cert-manager; do sleep 1; done'
-	$(GINKGO) -nodes 1 test/e2e/ -- --kubeconfig $$(pwd)/kubeconfig.yaml --project jetstack-cas --location europe-west1 --capoolid issuer-e2e
-	$(KIND) delete cluster --name casissuer-e2e
+e2e: $(BINDIR)/kind $(BINDIR)/kustomize $(BINDIR)/ginkgo $(BINDIR)/kubectl docker-build
+	./hack/ci/run-e2e.sh
```
Comment (JoshVanL) [inline Makefile:22]: Move to a test script

File: config/crd/bases/cas-issuer.jetstack.io_googlecasclusterissuers.yaml [hunk 3]
```diff
@@ -1,10 +1,9 @@
-
 ---
 apiVersion: apiextensions.k8s.io/v1
 kind: CustomResourceDefinition
 metadata:
   annotations:
-    controller-gen.kubebuilder.io/version: v0.5.0
+    controller-gen.kubebuilder.io/version: v0.9.2
```
Comment (JoshVanL) [inline config/crd/bases/cas-issuer.jetstack.io_googlecasclusterissuers.yaml:6]: From updated deps.

File: test/e2e/suite/issuers/issuers.go [hunk 4]
```diff
@@ -69,28 +64,20 @@ var _ = framework.CasesDescribe("issuers", func() {
 	f := framework.NewDefaultFramework("issuer")
 	cfg := config.GetConfig()
 	It("Tests Issuer functionality", func() {
-		By("Decrypting issuer secret")
-		encSecret := bytes.NewReader(googleSa)
-		identity, err := age.ParseX25519Identity(os.Getenv("AGE_SECRET_KEY"))
-		Expect(err).NotTo(HaveOccurred())
-		decSecret, err := age.Decrypt(encSecret, identity)
```
Comment (JoshVanL) [inline test/e2e/suite/issuers/issuers.go:76]: No longer using an age secret.

File: .github/workflows/tests.yml [hunk 5]
```diff
@@ -11,6 +11,9 @@ on:
     - cron: "23 3 */3 * *"
 jobs:
   run_tests:
+    permissions:
```
Comment (irbekrm) [inline .github/workflows/tests.yml]: Question: the token is needed for workload identity auth?

File: Makefile [hunk 6]
```diff
@@ -28,6 +33,10 @@ e2e: $(BINDIR)/kind $(BINDIR)/kustomize $(BINDIR)/ginkgo $(BINDIR)/kubectl docke
 	timeout 5m bash -c 'until $(KUBECTL) --kubeconfig kubeconfig.yaml --timeout=120s wait --for=condition=Ready pods --all --namespace kube-system; do sleep 1; done'
 	timeout 5m bash -c 'until $(KUBECTL) --kubeconfig kubeconfig.yaml --timeout=120s wait --for=condition=Ready pods --all --namespace cert-manager; do sleep 1; done'
 	$(GINKGO) -nodes 1 test/e2e/ -- --kubeconfig $$(pwd)/kubeconfig.yaml --project jetstack-cas --location europe-west1 --capoolid issuer-e2e
+	rm -rf ${E2E_LOG_DIR} && mkdir -p ${E2E_LOG_DIR}
+	$(KIND) export logs --name casissuer-e2e ${E2E_LOG_DIR}
+	$(KUBECTL) --kubeconfig kubeconfig.yaml cluster-info dump --all-namespaces --output-directory ${E2E_LOG_DIR}/kubectl-cluster-info-dump --output yaml
+	$(KUBECTL) --kubeconfig kubeconfig.yaml describe cert-manager --all-namespaces > ${E2E_LOG_DIR}/cert-manager-resources.yaml
```
Comment (irbekrm) [inline Makefile]: I didn't know you could do this 👍🏼

File: hack/ci/run-e2e.sh [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+kubectl --kubeconfig $KUBECONFIG apply -f https://github.com/jetstack/cert-manager/releases/download/v1.5.1/cert-manager.yaml
+kustomize build config/crd | kubectl --kubeconfig $KUBECONFIG apply -f -
+cd config/manager; kustomize edit set image controller=${IMG}
+kustomize build config/default | kubectl --kubeconfig $KUBECONFIG apply -f -
+timeout 5m bash -c 'until $(KUBECTL) --kubeconfig $KUBECONFIG --timeout=120s wait --for=condition=Ready pods --all --namespace kube-system; do sleep 1; done'
+timeout 5m bash -c 'until $(KUBECTL) --kubeconfig $KUBECONFIG --timeout=120s wait --for=condition=Ready pods --all --namespace cert-manager; do sleep 1; done'
+trap export_logs EXIT
+ginkgo -nodes 1 test/e2e/ -- --kubeconfig $$(pwd)/$KUBECONFIG --project jetstack-cas --location europe-west1 --capoolid issuer-e2e
+$(KUBECTL) --kubeconfig $KUBECONFIG cluster-info dump --all-namespaces --output-directory ${E2E_LOG_DIR}/kubectl-cluster-info-dump --output yaml
+$(KUBECTL) --kubeconfig $KUBECONFIG describe cert-manager --all-namespaces > ${E2E_LOG_DIR}/cert-manager-resources.yaml
```
Comment (irbekrm) [inline hack/ci/run-e2e.sh]: Question: why not put this into the `export_logs` function?

**Review by irbekrm — APPROVED** [review]
All look like good improvements to me, I've added a couple questions and nits, but happy for you to merge when ready 👍🏼 

/hold
/lgtm

File: Makefile [hunk 8]
```diff
@@ -91,7 +91,7 @@ $(BINDIR)/kind: $(BINDIR)
 KIND=$(BINDIR)/kind
 
 $(BINDIR)/ginkgo: $(BINDIR)
-	cd hack/tools && go build -o $@ github.com/onsi/ginkgo/v2/ginkgo
+	cd hack/tools && go build -o $@ github.com/onsi/ginkgo/ginkgo
```
[the remark this answers is not in the document]
Comment (JoshVanL) [reply Makefile:85]: Nothing. The project hasn't been updated to v2 yet.

File: .github/workflows/tests.yml (hunk 5 above)
Comment (JoshVanL) [reply .github/workflows/tests.yml]: Yes, exactly.

https://github.com/marketplace/actions/authenticate-to-google-cloud#usage

**Comment by JoshVanL** [comment]
Thanks for the review @irbekrm, please take another look

**Comment by irbekrm** [comment]
Thanks Josh!

/lgtm
/hold cancel

**Pull request merged**